src/System.Net.Mime/src/System/Net/Mail/MailAddress.cs

```diff
@@ -0,0 +1,260 @@
+            }
+        }
+
+        public string User
+        {
+            get
+            {
+                return _userName;
+            }
+        }
+
+        private string GetUser(bool allowUnicode)
+        {
+            // Unicode usernames cannot be downgraded
+            if (!allowUnicode && !MimeBasePart.IsAscii(_userName, true))
+            {
+                throw new SmtpException(SR.Format(SR.SmtpNonAsciiUserNotSupported, Address));
+            }
+            return _userName;
+        }
+
+        public string Host
+        {
+            get
+            {
+                return _host;
+            }
+        }
+
+        private string GetHost(bool allowUnicode)
+        {
+            string domain = _host;
+
+            // Downgrade Unicode domain names
+            if (!allowUnicode && !MimeBasePart.IsAscii(domain, true))
+            {
+                IdnMapping mapping = new IdnMapping();
+                try
+                {
+                    domain = mapping.GetAscii(domain);
+                }
+                catch (ArgumentException argEx)
+                {
+                    throw new SmtpException(SR.Format(SR.SmtpInvalidHostName, Address), argEx);
+                }
+            }
+            return domain;
+        }
+
+        public string Address
+        {
+            get
+            {
+                return String.Format(CultureInfo.InvariantCulture, "{0}@{1}", _userName, _host);
+            }
+        }
+
+        private string GetAddress(bool allowUnicode)
+        {
+            return String.Format(CultureInfo.InvariantCulture, "{0}@{1}",
+                GetUser(allowUnicode), GetHost(allowUnicode));
+        }
+
+        private string SmtpAddress
+        {
+            get
+            {
+                return String.Format(CultureInfo.InvariantCulture, "<{0}>", Address);
+            }
+        }
+
+        internal string GetSmtpAddress(bool allowUnicode)
+        {
+            return String.Format(CultureInfo.InvariantCulture, "<{0}>", GetAddress(allowUnicode));
+        }
+
+        /// <summary>
+        /// this returns the full address with quoted display name.
+        /// i.e. "some email address display name" <user@host>
+        /// if displayname is not provided then this returns only user@host (no angle brackets)
+        /// </summary>
+        /// <returns></returns>
+        public override string ToString()
+        {
+            if (String.IsNullOrEmpty(DisplayName))
+            {
+                return this.Address;
+            }
+            else
+            {
+                return String.Format("\"{0}\" {1}", DisplayName, SmtpAddress);
+            }
+        }
+
+        public override bool Equals(object value)
+        {
+            if (value == null)
+            {
+                return false;
+            }
+            return ToString().Equals(value.ToString(), StringComparison.InvariantCultureIgnoreCase);
```

Comment:
The email address can be a UTF-8 string with non-ascii characters in the local part of the address, reference [here](https://tools.ietf.org/html/rfc6531#section-3.2). In which case, ```lasst``` and ```laßt``` gives true under invariantcultureignorecase and false under ordinalignorecase